core/bin/snapshots_creator/src/main.rs

```diff
@@ -0,0 +1,251 @@
+use anyhow::Context as _;
+use prometheus_exporter::PrometheusExporterConfig;
+use tokio::sync::watch;
+use tokio::sync::watch::Receiver;
+use vise::{Gauge, Metrics};
+use zksync_config::configs::PrometheusConfig;
+use zksync_config::PostgresConfig;
+
+use zksync_dal::ConnectionPool;
+use zksync_env_config::object_store::SnapshotsObjectStoreConfig;
+use zksync_env_config::FromEnv;
+use zksync_object_store::{ObjectStore, ObjectStoreFactory};
+use zksync_types::snapshots::{
+    SnapshotFactoryDependencies, SnapshotStorageLogsChunk, SnapshotStorageLogsStorageKey,
+};
+use zksync_types::zkevm_test_harness::zk_evm::zkevm_opcode_defs::decoding::AllowedPcOrImm;
+use zksync_types::{L1BatchNumber, MiniblockNumber};
+use zksync_utils::ceil_div;
+use zksync_utils::time::seconds_since_epoch;
+
+#[derive(Debug, Metrics)]
+#[metrics(prefix = "snapshots_creator")]
+struct SnapshotsCreatorMetrics {
+    pub storage_logs_chunks_count: Gauge<u64>,
+
+    pub snapshot_generation_duration: Gauge<u64>,
+
+    pub snapshot_l1_batch: Gauge<u64>,
+
+    pub snapshot_generation_timestamp: Gauge<u64>,
+}
+#[vise::register]
+pub(crate) static METRICS: vise::Global<SnapshotsCreatorMetrics> = vise::Global::new();
+
+async fn maybe_enable_prometheus_metrics(stop_receiver: Receiver<bool>) -> anyhow::Result<()> {
+    let prometheus_config = PrometheusConfig::from_env().ok();
+    if let Some(prometheus_config) = prometheus_config {
+        let exporter_config = PrometheusExporterConfig::push(
+            prometheus_config.gateway_endpoint(),
+            prometheus_config.push_interval(),
+        );
+
+        tracing::info!("Starting prometheus exporter with config {prometheus_config:?}");
+        tokio::spawn(exporter_config.run(stop_receiver));
+    } else {
+        tracing::info!("Starting without prometheus exporter");
+    }
+    Ok(())
+}
+async fn process_storage_logs_single_chunk(
+    blob_store: &dyn ObjectStore,
+    pool: &ConnectionPool,
+    l1_batch_number: L1BatchNumber,
+    chunk_id: u64,
+    chunk_size: u64,
+    chunks_count: u64,
```

Comment:
Nit: Is this arg just used for logging? If so, I'd move logging to the caller site, so that passing this arg is not necessary.

Reply:
yeah, it's used only for that. I moved this logging back and forth at least twice yesterday. I totally agree that it's not ideal to have an argument just for that and it's a bit ankward, but I find the 'main logic' code a bit more readable when the logging is moved inside the method, so in the end I left it there